add summit packages-extended.yaml

bd4 opened the pull request.

bd4 commented:
WIP because I am testing the compiler assignment - xgc1 is still trying to use gcc instead of pgi for some reason. I guess this was also the case with the old config... so maybe it is still better to merge to get wider testing sooner.

germasch commented on spack/configs/summit/packages.yaml:
```diff
@@ -1,25 +1,61 @@
+
+  dataspaces:
+    version: [develop]
+    variants: +fpic maxdims=10 network-type=infiniband +dimes
```
So I'm about unsure about whether this is a good idea, because this is setting defaults that are global in my account even though these are (I think) specific to gene/xgc1.  But xgc1/gene already specify what they want from adios, so if prereqs for them already get installed correctly, and so the above isn't really used -- it's used if I install adios{,2] directly, and who knows what my preferences are for that ;)

(I'm less sure about dataspaces -- maybe some of those parameters are summit-specific, so in that case, it 'd make sense to have that in there.)

bd4 replied:
Good point, adios/adios2 shouldn't need to be there, I'll remove them.

germasch commented on spack/configs/summit/packages.yaml:
```diff
@@ -1,25 +1,61 @@
   fftw:
     paths:
       fftw@3.3.8%gcc@8.1.1: /sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/gcc-8.1.1/fftw-3.3.8-in2y7enjd3memwrmamnoxvsmx37mse7t
+      fftw@3.3.8%gcc@19.9: /sw/summit/.swci/1-compute/opt/spack/20180914/linux-rhel7-ppc64le/pgi-19.9/fftw-3.3.8-wri566ow7qlimailhyznou5cfm4h362z
```
That probably was meant to say `pgi@19.9`.

bd4 commented on spack/configs/summit/packages.yaml:
```diff
@@ -2,16 +2,27 @@ packages:
+    compiler: [gcc@8.1.1]
+
+  xgc1:
+    compiler: [pgi@19.9]
```
This isn't working for me, the default xgc1 spec is still to use `gcc@8.1.1`. This might be because a bunch of it's dependencies are already installed with gcc for gene. The 'which version should I use by default' question is a complicated one - for externals, I generally want it to heavily favor pre-installed versions, in many cases even if there is a compiler mismatch (which can be achieved by not specifiying a compiler for the external even when known, which is what I have been doing). In this case, it seems like it should be obvious - I explicitly said I want to build xgc1 with pgi@19.9 and nothing else, so all it's deps should follow.

germasch replied:
I have no idea why this doesn't work, either -- I'd have thought it should work. And it'd be useful to be able to set such defaults (though unless there's a specific reason not to, I'd have gone with the latest pgi).

bd4 commented:
I think it's useful to divide packages into categories, and then discuss whether they should be externals or not:

1. system: MPI and Cuda, very system dependent, often no benefit to installing (unless the system version is broken, *cough* longhorn)
2. core: immediate dependencies of gene and XGC - scalapack, lapack, blas, fftw, petsc, slepc. This can be further divided into whether there is a stable system module available
3. build tools: autotools, cmake, flex, bison
4. perf: adios, adios2 and all it's dependencies

I generally like packages.yaml to have (1) system, (2) core when system module available (e.g. netlib-scalapack module on summit), (3) build tools as long as they work (having newest version often not necessary). I have mixed feelings about (4), without some externals building takes forever, and I have run into a lot of spack build failures in adios2 deps when I try to install adios2. So curating this based on what we are able to get to work on the machine is useful. On the other hand, sometimes the externals don't work, so it's a lot of work to go through and curate this.

germasch commented:
> I think it's useful to divide packages into categories, and then discuss whether they should be externals or not:

I agree with this categorization.

> 1. system: MPI and Cuda, very system dependent, often no benefit to installing (unless the system version is broken, _cough_ longhorn)

Yeah, so those are things that one really should use the system-provided packages, because they should be optimized/system for the system anyway, and also because they'd be a pain to know how to configure (e.g., specific interconnect hardware, batch system, etc.) This more or less means using system-provided compilers, too, since at least for Fortran it's not really possible to mix MPI compiled with one compiler with an application compiled by another.

> 2. core: immediate dependencies of gene and XGC - scalapack, lapack, blas, fftw, petsc, slepc. This can be further divided into whether there is a stable system module available
> 3. build tools: autotools, cmake, flex, bison
> 4. perf: adios, adios2 and all it's dependencies

So I think there only fairly few packages where it's clear that the system provided packages will do the job (as opposed to giving other problems when compiling dependents). blas, lapack should be good, and fftw and scalapack as well as long as they're compiled with the correct mpi. System-provided petsc and slepc are rarely actually useful, because they need to be the right version of petsc (they break compatibility a lot), they need to include the correct features (e.g., suplerlu_dist for xgc), and need to match the correct scalar type (e.g. complex double for gene).

build tools etc: I definitely don't think they need to be compiled with a matching compiler (in fact that's one of my gripes with spack). Mostly, I'd agree that system-installed ones should work, though having a recent enough version of cmake can be a bit hit-or-miss.

In terms of adios2, at least, I also find that the system installed ones are often not what I need, being too out of date, or missing the python bindings, or whatever.

Also on that list, by the way: python.

In any case, I think there's nothing wrong with listing everything in 2. - 4. in a full packages.yaml, with the caveat that I don't fully understand how spack decides on what to prefer when it concretizes things. Those packages, I'd say, should mostly behave as options, ie., if there happens to be a petsc with the right version etc, the sytem one should be picked, but if one uses a different compiler/mpi/different variants, a custom one should be built.

I think there are pros and cons for both the basic and the full packages.yaml. I'd say the reason that I personally prefer to basic one (for now, anyway), is that it removes a lot of the dependency on machine-specific properties. Ie., I think the place where things often go wrong is where one mixes and matches things, some things that have already been manually or system installed with things one is building oneself. If you build pretty much everything yourself, it means it'll look the same on all machines that you're running on, and spack definitely has its quirks, but usually after you fix things at the spack level on one machine, compiling the same stack on the next machine then works smoothly.


> 
> I generally like packages.yaml to have (1) system, (2) core when system module available (e.g. netlib-scalapack module on summit), (3) build tools as long as they work (having newest version often not necessary). I have mixed feelings about (4), without some externals building takes forever, and I have run into a lot of spack build failures in adios2 deps when I try to install adios2. So curating this based on what we are able to get to work on the machine is useful. On the other hand, sometimes the externals don't work, so it's a lot of work to go through and curate this.

bd4 commented:
Ok so for each machine, we provide at least a minimal `packages.yaml` with MPI, Cuda, and any of scalapack, fftw, blas, etc that have well supported compiler specific system modules (right now only summit, hopefully the other new exa machines will too). This is the default one mentioned in the docs.

We can also provide a `packages-extended.yaml` (open to name suggestions) which can have anything and everything, and users can use it as a starting point to trim it down to what they really care about. Basically it serves as a reference for what is available on the system. The `spack-gen-external-packages.py` script can be used to generate an initial version of this. For now I think the core `packages.yaml` needs to be hand written for necessary system modules/compiler combos - the output of this script can be appended for the extend version. A note in the docs can mention that this version exists for some machines.

@germasch what do you think?

bd4 commented:
I reworked this to include the changes in `packages-extended.yaml` rather than changing the better tested `packages.yaml`. I did make a few changes to the main version - reformatting, adding pgi externals, fixed openssl version.

There are a few entries in -extended which may be worth copying into the main version - like automake, autoconf, bison, etc. I'll leave that for a separate discussion / PR though, this PR has sat around for too long.

germasch commented:
Looks good, thanks!
